discover: move instance discover code into separate package

magiconair opened the pull request:
This patch moves the code that discovers instances from metadata
information into a separate package with sub-packages for each
provider. Additionally, the dependencies for each provider are
vendored in separately to make the sub packages independent
from another.

magiconair commented:
I can drop the provider prefix in the filename again. I tried that for readability but it should be OK without. 

As for vendoring: you're suggesting that we have one `discover` package with one set of dependencies. `govendor` pulls in sub-dependencies and does not flatten them. I've tried that. Therefore, if consul-enterprise needs the S3 lib it will have to have its own copy of the library anyway since it cannot use the one from the discover package.

The benefit of the sub-vendoring is though that you can pick and choose which providers you want to support and only get their deps.

magiconair commented:
I think we should skip the step of the internal `github.com/hashicorp/consul/discover` package since the main motivation was to make discovery accessible to nomad. I suggest to move the code to a `github.com/hashicorp/discover` repository. Then I'd change the PR to swap out the current code for the lib.

@dadgar What is your take?

slackpad commented:
> As for vendoring: you're suggesting that we have one discover package with one set of dependencies. govendor pulls in sub-dependencies and does not flatten them. I've tried that. Therefore, if consul-enterprise needs the S3 lib it will have to have its own copy of the library anyway since it cannot use the one from the discover package.

Not that. Keeping each discover implementation in its own package should be all we need. We definitely don't want just a single flat discover package. That way if there are a dozen packages under discover and consul pulls in just  three (discover/aws, discover/gce, discover/azure) then govendor will only include the deps for those three.

magiconair commented:
After discussing this internally we've decided to create a github.com/hashicorp/go-discover package and move the code there as is. This includes vendoring the dependencies per provider and not for the repository as a whole. 

In addition, I'll add a generic `Discover` function which takes a single string parameter which carries the parameters for the different cloud providers. 

Format suggestion:

A single line string of `key=value` pairs where the value is url encoded, i.e. use `+` to encode a space and `%3D` for a `=`

```
provider=gce   project=v zonepattern=v tagvalue=v credfile=v
provider=aws   region=v  tagkey=v      tagvalue=v accesskey=v secretkey=v
provider=azure tagname=v tagvalue=v    clientid=v tenantid=v  secretkey=v subscriptionid=v
```

magiconair commented:
The only step missing is to move the github.com/hashicorp/consul/discover package to the github.com/hashicorp/discover repo.

dadgar commented on discover/discover.go:
```diff
@@ -0,0 +1,85 @@
+	"github.com/hashicorp/discover/gce"
+)
+
+func Discover(cfg string, l *log.Logger) ([]string, error) {
```
May want to consider making discoverer an interface an having a list of them that this default Discover method will lookup against.

You can see the pattern here: https://github.com/hashicorp/go-getter/blob/master/get.go#L25-L63.

This has several benefits, projects can add discoverers without upstreaming them, projects can disable discoverers, parsing of configs can live with the discoverer, etc.

magiconair replied:
I think interfaces should be defined with the consumer and not the producer. If someone needs a `Discoverer` interface then they can easily define it. 

Another issue is that although the signature of the `(aws|gce|azure).Discover` methods are similar they differ in their config structures. Therefore, the `discover.Discover` method is the generalized function. We might want to consider passing either a string and provide a parser or to pass a `map[string]string` to make the signature truly generic.

dadgar replied:
I would do the map[string]string and not sure I follow on the interface argument. The go-getter style is pretty nice since you can manipulate the list of available getters and still use all the built in methods of the library. And by having a list you can append to it an init function to make a new type available. Otherwise you end up having to wrap the library.

magiconair replied:
@dadgar ah, you want a registry. Sure, that could make sense.

magiconair commented:
@dadgar pls have another look. I've defined a `Discover` function type instead of an interface since that is what we need right now and if we were to change it later we would break some code one way or another.

dadgar commented:
@magiconair LGTM! That is what I had in mind!

magiconair commented:
OK, I'll give the comments and docs some more review time tomorrow and then move this to github.com/hashicorp/go-discover.

magiconair commented:
@dadgar I've dropped the `map[string]string` for a simple `string` and moved the `Parse` function into a separate package. This way we could change or support additional formats in the future but the main reason was that I didn't see the `map[string]string` signature in the stdlib. If it is generic enough to be passed as strings then we might as well just pass a single string. 

I've also updated the comments to include information I've previously dropped.
